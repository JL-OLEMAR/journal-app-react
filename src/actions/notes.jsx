import { collection, deleteDoc, addDoc, updateDoc, doc } from 'firebase/firestore'
import Swal from 'sweetalert2/dist/sweetalert2.js'
import 'sweetalert2/dist/sweetalert2.css'

import { db } from '../firebase/firebase-config.jsx'
import { fileUpload } from '../helpers/fileUpload.jsx'
import { loadNotes } from '../helpers/loadNotes.jsx'
import { types } from '../types/types.jsx'

// Note active(selected)
export const activeNote = (id, note) => ({
  type: types.notesActive,
  payload: { id, ...note }
})

// Add new note
export const addNewNote = (id, note) => ({
  type: types.notesAddNew,
  payload: { id, ...note }
})

// Add new note to firestore
export const startNewNote = () => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth

    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime() // in miliseconds
    }

    try {
      const doc = await addDoc(collection(db, `${uid}/journal/notes`), newNote)

      dispatch(addNewNote(doc.id, newNote))
      dispatch(activeNote(doc.id, newNote))
    } catch (error) {
      console.log(error)
    }
  }
}

// Load notes
export const setNotes = (notes) => ({
  type: types.notesLoad,
  payload: notes
})

// Load notes Firebase
export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    try {
      const notes = await (await loadNotes(uid))

      // sort notes by date desc
      const orderedNotes = notes.sort((a, b) => b.date - a.date)

      dispatch(setNotes(orderedNotes))
    } catch (error) {
      console.log(error)
    }
  }
}

// Reload a note with your fields save or updated
export const refreshNote = (id, note) => ({
  type: types.notesUpdated,
  payload: {
    id,
    note: { id, ...note }
  }
})

// Save or update a note to Firebase
export const startSaveNote = (note) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth

    !note.url && delete note.url

    try {
      const noteToFirestore = {
        ...note,
        date: new Date().getTime()
      }

      delete noteToFirestore.id
      await updateDoc(doc(db, `${uid}/journal/notes`, `${note.id}`), noteToFirestore)

      dispatch(refreshNote(note.id, noteToFirestore))
      Swal.fire('Note Saved', note.title, 'success')
    } catch (error) {
      console.log(error)
    }
  }
}

// Upload image Firebase
export const startUploading = (file) => {
  return async (dispatch, getState) => {
    const { active: activeNote } = getState().notes

    Swal.fire({
      title: 'Uploading...',
      text: 'Please wait...',
      allowOutsideClick: false,
      didOpen: () => Swal.showLoading()
    })

    const fileUrl = await fileUpload(file)

    // Set the name of the url to the selected note
    activeNote.url = fileUrl
    dispatch(startSaveNote(activeNote))
    Swal.close()
  }
}

// Delete note Firebase
export const startDeleting = (id) => {
  return async (dispatch, getState) => {
    try {
      const { uid, name } = getState().auth

      await deleteDoc(doc(db, `${uid}/journal/notes`, `${id}`))
      Swal.fire({
        title: `${name}, Are you sure?`,
        text: "You won't be able to revert this!",
        icon: 'warning',
        showCancelButton: true,
        confirmButtonColor: '#d33',
        cancelButtonColor: '#3085d6',
        confirmButtonText: 'Yes, delete it!'
      }).then((result) => {
        if (result.isConfirmed) {
          Swal.fire(
            'Deleted!',
            'Your note has been deleted.',
            'success'
          )
          dispatch(deleteNote(id))
        }
      })
    } catch (error) {
      console.log(error)
    }
  }
}

// Delete note
export const deleteNote = (id) => ({
  type: types.notesDelete,
  payload: id
})

// Clear note
export const noteLogout = () => ({
  type: types.notesLogoutCleaning
})
