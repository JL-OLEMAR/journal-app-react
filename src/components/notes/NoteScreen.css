.notes-main-content {
  display: flex;
  flex-direction: column;
  height: 100%;

  & .notes-content {
    background-color: #fff;
    border-radius: 3px;
    box-shadow: 0 3px #363636;
    display: flex;
    flex-direction: column;
    height: 100%;
    padding: 20px;

    & .title-input {
      border-radius: 5px;
      border: 1px solid #d8d8d8;
      color: #363636;
      font-size: 25px;
      font-weight: 700;
      margin-bottom: 10px;
      padding: 3px 6px;

      &:focus {
        outline: none;
      }
    }

    & .text-area {
      border-radius: 5px;
      border: 1px solid #d8d8d8;
      color: #333;
      cursor: auto;
      flex: 1 1 auto;
      font-size: 18px;
      margin-bottom: 10px;
      padding: 6px;
      resize: none;

      &:focus {
        outline: none;
      }

      &::-webkit-scrollbar {
        width: 10px;
      }

      &::-webkit-scrollbar-track {
        background-color: #e6e6e6;
        border-radius: 50px;
      }

      &::-webkit-scrollbar-thumb {
        background-color: #7e7e7e;
        border-radius: 50px;
      }
    }

    & .notes-image {
      & img {
        box-shadow: 5px 5px #363636;
        height: 150px;
      }
    }
  }
}
