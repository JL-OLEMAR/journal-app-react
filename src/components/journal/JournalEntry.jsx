import { useDispatch } from 'react-redux'
import moment from 'moment'

import { activeNote } from '../../actions/notes.jsx'

export const JournalEntry = ({ id, date, title, body, url }) => {
  const noteDate = moment(date)
  const dispatch = useDispatch()

  const handleEntryCLick = () => {
    dispatch(activeNote(id, { date, title, body, url }))
  }

  return (
    <div
      className='journal-entry pointer animate__animated animate__fadeIn animate__faster'
      onClick={handleEntryCLick}
    >
      {
        url && (
          <div
            className='entry-picture'
            style={{
              backgroundSize: 'cover',
              backgroundImage: `url(${url})`
            }}
          />)
      }

      <div className='entry-body'>
        <p className='entry-title'>
          {title}
        </p>
      </div>

      <div className='entry-date-box'>
        <span>{noteDate.format('dddd')}</span>
        <h4>{noteDate.format('Do')}</h4>
      </div>
    </div>
  )
}
