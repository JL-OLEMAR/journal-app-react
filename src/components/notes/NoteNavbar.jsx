import { useDispatch, useSelector } from 'react-redux'

import { startSaveNote, startUploading } from '../../actions/notes.jsx'

export const NoteNavbar = () => {
  const dispatch = useDispatch()
  const { active } = useSelector(state => state.notes)

  const handleSave = () => {
    dispatch(startSaveNote(active))
  }

  const handlePictureCLick = () => {
    document.querySelector('#fileSelector').click()
  }

  const handleFileChange = (e) => {
    const file = e.target.files[0]

    if (file) {
      dispatch(startUploading(file))
    }
  }

  return (
    <div className='notes-navbar'>
      <span> 02 de marzo 2022</span>

      <input
        id='fileSelector'
        name='file'
        style={{ display: 'none' }}
        type='file'
        onChange={handleFileChange}
      />

      <div>
        <button
          className='btn btn-picture'
          onClick={handlePictureCLick}
        >
          <i className='far fa-image fa-1x mt-1 pr-1' />
          <span>Picture</span>
        </button>

        <button
          className='btn btn-save'
          onClick={handleSave}
        >
          <i className='far fa-save fa-1x mt-1 pr-1' />
          <span>Save</span>
        </button>
      </div>
    </div>
  )
}
