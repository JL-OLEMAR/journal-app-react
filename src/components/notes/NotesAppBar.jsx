import { useDispatch, useSelector } from 'react-redux'

import { startSaveNote, startUploading } from '../../actions/notes.jsx'

export const NotesAppBar = () => {
  const dispatch = useDispatch()
  const { active } = useSelector(state => state.notes)

  const handleSave = () => {
    dispatch(startSaveNote(active))
  }

  const handlePictureCLick = () => {
    document.querySelector('#fileSelector').click()
  }

  const handleFileChange = (e) => {
    const file = e.target.files[0]

    if (file) {
      dispatch(startUploading(file))
    }
  }

  return (
    <div className='notes-appbar'>
      <span> 28 de agosto 2021</span>

      <input
        id='fileSelector'
        name='file'
        style={{ display: 'none' }}
        type='file'
        onChange={handleFileChange}
      />

      <div>
        <button
          className='btn'
          onClick={handlePictureCLick}
        >
          Picture
        </button>

        <button
          className='btn'
          onClick={handleSave}
        >
          Save
        </button>
      </div>
    </div>
  )
}
