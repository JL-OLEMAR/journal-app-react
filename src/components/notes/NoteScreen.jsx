import { useEffect, useRef } from 'react'
import { useDispatch, useSelector } from 'react-redux'

import { useForm } from '../../hooks/useForm.jsx'
import { activeNote, startDeleting } from '../../actions/notes.jsx'

import { NotesAppBar } from './NotesAppBar.jsx'
import './NoteScreen.css'

export const NoteScreen = () => {
  const dispatch = useDispatch()
  const { active: note } = useSelector((state) => state.notes)
  const [formValues, handleInputChange, reset] = useForm(note)
  const { body, title, id } = formValues
  const activeId = useRef(note.id)

  useEffect(() => {
    if (note.id !== activeId.current) {
      reset(note)
      activeId.current = note.id
    }
  }, [note, reset])

  useEffect(() => {
    dispatch(activeNote(formValues.id, { ...formValues }))
  }, [dispatch, formValues])

  const handleDelete = () => {
    dispatch(startDeleting(id))
  }

  return (
    <div className='notes-main-content'>
      <NotesAppBar />

      <div className='notes-content'>
        <input
          autoComplete='off'
          className='title-input'
          name='title'
          placeholder='Some awesome title'
          type='text'
          value={title}
          onChange={handleInputChange}
        />

        <textarea
          className='text-area'
          name='body'
          placeholder='What happend today'
          value={body}
          onChange={handleInputChange}
        />

        {note.url && (
          <div className='notes-image'>
            <img alt={note.title} src={note.url} />
          </div>
        )}
      </div>

      <button className='btn btn-danger' onClick={handleDelete}>
        Delete
      </button>
    </div>
  )
}
