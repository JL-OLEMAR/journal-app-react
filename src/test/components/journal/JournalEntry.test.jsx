/**
 * @jest-environment node
 */

import { describe, expect, test, vi } from 'vitest'
import { MemoryRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import { mount } from 'enzyme'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'

import { activeNote } from '../../../actions/notes.jsx'
import { JournalEntry } from '../../../components/journal/JournalEntry.jsx'
import 'jsdom-global/register' // → For mount a component JournalEntry

const middlewares = [thunk]
const mockStore = configureStore(middlewares)
const initState = {}
const store = mockStore(initState) // → mock store

store.dispatch = vi.fn() // → Mock dispatch

// Props of JournalEntry
const note = {
  id: 10,
  date: 0,
  title: 'Hello',
  body: 'World',
  url: 'https://somewhere.com/photo.jpg'
}

const wrapper = mount(
  <Provider store={store}>
    <MemoryRouter>
      <JournalEntry {...note} />
    </MemoryRouter>
  </Provider>
)

describe('Tests <JournalEntry />', () => {
  test("should show correctly 'snapshot'", () => {
    expect(wrapper).toMatchSnapshot()
  })

  test('should note active', () => {
    wrapper.find('div.journal-entry').prop('onClick')()
    expect(store.dispatch).toHaveBeenCalledWith(
      activeNote(note.id, { ...note }) // send note to activeNote action
    )
  })
})
