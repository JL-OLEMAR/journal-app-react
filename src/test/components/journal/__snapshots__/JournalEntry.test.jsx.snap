// Vitest Snapshot v1

exports[`Tests <JournalEntry /> > should show correctly 'snapshot' 1`] = `
<a
  className="ancla-note"
  href="/#note"
  onClick={[Function]}
>
  <div
    className="journal-entry pointer animate__animated animate__fadeIn animate__faster"
    onClick={[Function]}
  >
    <div
      className="entry-picture"
      style={
        {
          "backgroundImage": "url(https://somewhere.com/photo.jpg)",
          "backgroundSize": "cover",
        }
      }
    />
    <div
      className="entry-body"
    >
      <p
        className="entry-title"
      >
        Hello
      </p>
    </div>
    <div
      className="entry-date-box config-center"
    >
      <span>
        Wednesday
      </span>
      <h5>
        Dec
        /
        31
      </h5>
    </div>
  </div>
</a>
`;
